### ISSUE TITLE
Load developer stencil directory dosn't work

### ISSUE THREAD

**Issue by spider4216** [evolus/pencil#758]
Hi there! First of all I want to say thank you for great tool for make GUI prototypes! I use it quite long and it's very helpful for my projects!
Recently I decided to create collection for prototyping GUI int the first time but faced with problem

According to manual Pencil provides a convenient way to load and reload stencil collection that is under development via "Load developer stencil directory" feature but it doesn't work for my project.

I created directory structure:

```
- icons/
- - - - Header-Purple.png
- Definition.xml 
```
My xml contains:
```
<Shapes xmlns="http://www.evolus.vn/Namespace/Pencil"
    xmlns:p="http://www.evolus.vn/Namespace/Pencil"
    xmlns:svg="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
    id="ysirotenko-pencil-collection-1"
    displayName="KaIOS Components"
    description="Components for design apps for mobile operation system KaiOS"
    author="Yuriy Sirotenko"
    url="https://ysirotenko.com">
<!--Your shapes go here -->
    
    <Shape id="header" displayName="Header" icon="icons/Header-Purple.png">
        <p:Content xmlns:p="http://www.evolus.vn/Namespace/Pencil" xmlns="http://www.w3.org/2000/svg">
            <path id="path" fill="none" style="stroke-linejoin: round;"/>
        </p:Content>
    </Shape>


</Shapes>
```
Then I specified directory where my "Definition.xml" was placed via "Load developer stencil directory" feature

![image](https://user-images.githubusercontent.com/10693428/212797993-11652d2a-1689-4163-8971-ee03bbbb4737.png)

but there isn't any my collection in panel:

![image](https://user-images.githubusercontent.com/10693428/212798111-0c4281bf-d3cf-4932-88ee-8b3fbf194ecd.png)

I pressed "F5", and saw popup with message "Developer collections were reloaded" but didn't see my collection

But when I zip this directory and install it from file, it's okay

![image](https://user-images.githubusercontent.com/10693428/212798208-b530b561-7a4f-4228-9ed5-5b37d77ae1fd.png)
![image](https://user-images.githubusercontent.com/10693428/212798256-eb12c596-3ee5-4091-bfa1-3a17afa101a3.png)

Actually, when I clicked "Load developer stencil directory" menu's item, error was appeared in log:

```
[251447:0117/084430.342128:ERROR:browser_main_loop.cc(269)] Gtk: gtk_file_chooser_widget_set_current_name: assertion 'priv->action == GTK_FILE_CHOOSER_ACTION_SAVE || priv->action == GTK_FILE_CHOOSER_ACTION_CREATE_FOLDER' failed

```

But everything is ok in Pencil GUI

Did I do something wrong? Thank you!

* Operating system: Ubuntu 20.04.5 LTS (Focal Fossa)
* Pencil version: 3.1.1
* Node v19.4.0
* NPM 9.2.0

Pencil was installed via source (npm install and npm start)

**Comment by donPablo2** [comment]
Tools/DeveloperTools/**LoadDeveloperStencilDirectory**…/  gives popup titles Select Developer Stencil ‘Definition.xml’ file
In menu it sez Directory, but on popup it sez File.  Also, in popup the filter allows only for Definition.xml(*.xml) 

After I select   Definition.xml and press “Open”, I do not see my new shape in any menu listing.

Windows Edition:	Windows 10 Pro   Version 21H2   OS build 19044.2486
Pencil version:  3.1.1

**Comment by rozek** [comment]
I just encountered the same problem: on my Mac, I cannot even open the *directory* but always have to select a file - which seems to be wrong according to [the docs](http://pencil.evolus.vn/wiki/devguide/Prepare_Dev_Env.html)

**Comment by spider4216** [comment]
I'm creating new stencil and have to build zip every time to see changes. It take too much time for developing...

**Comment by donPablo2** [comment]
yes, building a zip takes time.  Last week I discovered a "loophole"
!!  the myQuad folder does not need to be ZIP-ed;   just rename with suffix  .zip

I can open the Definition in notepad++, do edits, and save (not exit)
and then test via click taskbar Pencil.exe icon  (or click VSCode icon & F5)
and quit Pencil, come back and do more edits, save, test, etc.
```
``C:\Users\myUSERID\.pencil\stencils > tree    /f
 └───myQuad.zip
    │     Definition.xml
    └───Icons
          plain-text.png
```

**Comment by ghost** [comment]
Workaround:
1. Find your `config.json` file in Pencil configuration folder (on Linux, it is `~/.pencil/config.json`, probably the same on MacOS, no idea where it is on Windows)
2. Find this line: 
```json
    "dev.stencil.path": "none",
```
3. Set the value to the full path to your `Defintion.xml` file
4. Save the file
5. Open Pencil
6. Enjoy development

There is also a `dev.stencil.dir` which had no effect when modifying it.

Hint: if a software works on Linux, you can be sure there is a configuration file (on Linux, [everything is a file](https://en.wikipedia.org/wiki/Everything_is_a_file)). Even on Windows (rather than using Registry).

EDIT:
Actually, this setting can also be found in the advanced settings, right in Pencil's GUI.

**Comment by donPab** [comment]
@Arcesilas -- thanks for the workaround, 
and for bringing me back to look at Pencil.

I believe that the problem is in the usage of show.OpenDialog.

showOpenDialog is a part of electron.dialog.showOpenDialog in 
\app\node_modules\electron\dist\electron.exe

I believe that the Pencil v.3 upgrade fixed 
only 5 of the 17 usages.
From  }, function (filenames) {
To    }).then(result => {

I instrumented  CollectionManager.selectDeveloperStencilDir   code with—

>     dialog.showOpenDialog({
        title: "Select Developer Stencil 'Definition.xml' file",
        defaultPath: Config.get("dev.stencil.path") || os.homedir(),
        filters: [{ name: "Definition.xml", extensions: ["xml"] }]
    }).then(result => {
        console.log("sod-canceled=" + result.canceled)
        console.log("sod-paths>>>" + result.filePaths)
      }).catch(err => {console.log("sod-error>" + err) })
And I get console—  “sod” is shorthand for  “showOpenDialog”
sod-canceled=false 
sod-paths>>>C:\H_Paul\Pencil\$pg\pgBezier\Definition.xml<

It seems that the >>.then<< works, and that the >>,function<< fails
Only 5 of 17 of show.OpenDialog  usages have >.then< :
-->   DocumentHandler, collectionManager-1st, imageData, ExportDialog, FontDetailDialog
See these 5 for their implementation of    >then<.
The other 12 need to be fixed.

**Comment by ghost** [comment]
Actually, I have not looked at the entire code of Pencil. I've tried to run it from repository, but could not make it launch completely: the Pencil logo in the splash screen keeps waving/blinking (whatever verb is appropriate, I don't know how to tell). I've tried to find out where it stops, but I've never worked with Electron (I'm from the web, backend, in the PHP world). I have used and abused of console.log to try to figure out where the problem comes from, with no result.
I would love to be able to debug this to have it running and maybe be able to fix some issues, but without a working development version, it's a dead end for me.
